internal/js/console.go

```diff
@@ -110,62 +110,186 @@ func (c console) valueString(v sobek.Value) string {
 		return functionLog
 	}
-	// Skip non-object values.
+
+	// Handle non-object values.
 	obj, ok := v.(*sobek.Object)
 	if !ok {
-		return v
+		return formatPrimitive(v)
 	}
+
 	// Prevent circular references.
 	if seen[obj] {
 		return circularLog
 	}
 	seen[obj] = true
 	defer delete(seen, obj)
 
+	if obj.ClassName() == "Error" {
+		return v.String()
+	}
+
+	// Check for TypedArray and ArrayBuffer.
+	if isBinaryData(obj) {
+		return formatBinaryData(obj)
+	}
+
 	// Handle arrays element-by-element, recursively.
 	if obj.ClassName() == "Array" {
 		length := obj.Get("length").ToInteger()
-		arr := make([]any, length)
+		if length == 0 {
+			return "[]"
+		}
+		elements := make([]string, length)
 		for i := range length {
-			arr[i] = c.traverseValue(obj.Get(strconv.FormatInt(i, 10)), seen)
+			elements[i] = c.traverseValue(obj.Get(strconv.FormatInt(i, 10)), seen)
 		}
-		return arr
+		return "[ " + strings.Join(elements, ", ") + " ]"
 	}
 
 	keys := obj.Keys()
-	// Fast path for empty objects and other JS types with no enumerable properties.
+	// for empty objects and other JS types with no enumerable properties.
 	if len(keys) == 0 {
-		return obj
+		// Date objects have no enumerable keys but should display as ISO string.
+		if obj.ClassName() == "Date" {
+			return formatDate(obj)
+		}
+		return "{}"
 	}
+
 	// Handle objects key-by-key, recursively.
-	m := make(map[string]any, len(keys))
-	for _, key := range keys {
-		m[key] = c.traverseValue(obj.Get(key), seen)
+	pairs := make([]string, len(keys))
+	for i, key := range keys {
+		pairs[i] = key + ": " + c.traverseValue(obj.Get(key), seen)
+	}
+	return "{ " + strings.Join(pairs, ", ") + " }"
+}
+
+// formatDate formats a Date object as a quoted ISO string.
+func formatDate(obj *sobek.Object) string {
+	if toISOString, ok := sobek.AssertFunction(obj.Get("toISOString")); ok {
+		if result, err := toISOString(obj); err == nil {
+			return fmt.Sprintf("%q", result.String())
+		}
+	}
+	return "{}"
+}
+
+// formatPrimitive formats a primitive JS value as a string.
+func formatPrimitive(v sobek.Value) string {
+	switch v.ExportType().Kind() {
+	case reflect.String:
+		return fmt.Sprintf("%q", v.String())
+	default:
+		return v.String()
+	}
+}
+
+// checks for TypedArray and Array Buffer
+func isBinaryData(obj *sobek.Object) bool {
+	exportType := obj.ExportType()
+	if exportType == nil {
+		return false
+	}
+
+	if _, ok := obj.Export().(sobek.ArrayBuffer); ok {
+		return true
 	}
 
-	return m
+	if exportType.Kind() != reflect.Slice {
+		return false
+	}
+	switch exportType.Elem().Kind() {
+	case reflect.Int8, reflect.Uint8,
+		reflect.Int16, reflect.Uint16,
+		reflect.Int32, reflect.Uint32,
+		reflect.Float32, reflect.Float64,
+		reflect.Int64, reflect.Uint64:
+		return true
+	default:
+		return false
+	}
+}
+
+func formatBinaryData(obj *sobek.Object) string {
+	// ArrayBuffer
+	if ab, ok := obj.Export().(sobek.ArrayBuffer); ok {
+		bytes := ab.Bytes()
```

Comment:
I might be missing something, but wouldn't this be a good candidate to be refactored to use the stdlib's hex encoding package: https://pkg.go.dev/encoding/hex#Encode?

Reply:
Thanks for the clarification, your approach makes sense as I do prefer the current implementation's output 🙇‍♂️